framework/include/nodalkernels/ADNodalKernel.h

```diff
@@ -0,0 +1,58 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "NodalKernelBase.h"
+
+/**
+ * Base class for creating nodal kernels with AD-computed Jacobians
+ */
+class ADNodalKernel : public NodalKernelBase
+{
+public:
+  /**
+   * Class constructor.
+   * @param parameters The InputParameters for the object
+   */
+  static InputParameters validParams();
+
+  ADNodalKernel(const InputParameters & parameters);
+
+  /**
+   * Compute the residual at the current node.
+   *
+   * Note: This is NOT what a user would normally want to override.
+   * Usually a user would override computeQpResidual()
+   */
+  void computeResidual() override;
+
+  /**
+   * Compute the Jacobian at one node.
+   *
+   * Note: This is NOT what a user would normally want to override.
+   * Usually a user would override computeQpJacobian()
+   */
+  void computeJacobian() override;
+
+  /**
+   * This method simply routes to computeJacobian whenever jvar == _var.number() since global AD
+   * computes all the derivatives for all variables at once
+   */
+  void computeOffDiagJacobian(unsigned int jvar) override;
```

Comment:
If you're really trying to lock a user out of these I would `override final`

Reply:
If that's the case, they need to be virtual

Reply:
Nope. As soon as its marked `virtual` in a base class it's virtual all the way up the chain of derived classes. Moreover by marking something as `override` it's immediately explicit that the function is `virtual`. `override` on a non-virtual would be a compiler error

Reply:
Ah yeah, I found it in the standard. I suppose if you allow implicit virtual, you remove the implicit virtual with final